[MRG] Fix fetch_california_housing

lesteve opened the pull request:
Fix #5953.

I used http://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.tgz.

I still has an old cal_housing.pkz (i.e. before lib.stat.cmu.edu got broken)  in ~/scikit_learn_data so I double-checked that the content of the pickles were matching exactly.

Here are the plots from examples/ensemble/plot_partial_dependence.py:
![figure_1](https://cloud.githubusercontent.com/assets/1680079/11844040/edd15922-a40c-11e5-8c72-b819e7689c53.png)

![figure_2](https://cloud.githubusercontent.com/assets/1680079/11844049/f519ccaa-a40c-11e5-8ca1-b26ff607df87.png)

They match the ones from the [stable doc](http://scikit-learn.org/stable/auto_examples/ensemble/plot_partial_dependence.html)

![](http://scikit-learn.org/stable/_images/plot_partial_dependence_001.png)
![](http://scikit-learn.org/stable/_images/plot_partial_dependence_002.png)

lesteve committed:
FIX fetch_california_housing

Using a different resource than lib.stat.cmu.edu which has not been fixed yet.

lesteve committed:
Revert hotfixes to plot_partial_dependence.py

since fetch_california_housing is now fixed.

betatim commented:
Images look pixel perfect to me even with a :mag: 

* Does it make sense to remove the `main()` in the example? It seems more consistent with other examples but makes it harder to see what really changed. More a question of style/convention.

* The new download mechanism leaves several files and directories around my data directory in addition to the pickle. Would be good to clean those up, or avoid making them. Maybe wrap the downloading, unpacking, etc in a `try: ... finally:`. An alternative would be to pass a file object to `tarfile.open()` just like was done for `zipfile` to avoid creating the intermediate files.

lesteve commented:
> Does it make sense to remove the main() in the example? It seems more consistent with other examples but makes it harder to see what really changed.

The main was added as part of the hotfix https://github.com/scikit-learn/scikit-learn/commit/37c8d0297be0f63cf5ceea55259eed3629023885, not sure exactly why, maybe with the hope that `__name__` wouldn't be `"__main__"` when the example was run as part of the documentation.

> The new download mechanism leaves several files and directories around my data directory in addition to the pickle. Would be good to clean those up, or avoid making them. Maybe wrap the downloading, unpacking, etc in a try: ... finally:. An alternative would be to pass a file object to tarfile.open() just like was done for zipfile to avoid creating the intermediate files.

I have to admit I am not sure about the general conventions for scikit-learn downloaders and whether they should leave files around or not but I am fine with cleaning up if that's the consensual way to do it.

betatim commented:
Ok, if it was added as a part of a fix then it makes perfect sense to remove it.

Also unsure about the consensus about tidying up, it just feels like the right thing to do (or using `BytesIO` to avoid it).

ogrisel commented:
+1 `BytesIO` as this is a small dataset that easily fit in memory.

ogrisel commented:
The `__main__` block is needed to use multiprocessing under windows. Please restore the previous layout (and add a comment to explain).

Other than that +1 for merge.

lesteve committed:
Reinstate __name__ == __main__ check

so that the example can be run under Windows since plot_partial_dependence call uses multiprocessing

lesteve committed:
Do not leave files around after untarring the archive

ogrisel commented on sklearn/datasets/california_housing.py:
```diff
@@ -89,18 +90,19 @@ def fetch_california_housing(data_home=None, download_if_missing=True):
+        print('downloading Cal. housing from %s to %s' % (DATA_URL, archive_path))
+        urllib.urlretrieve(DATA_URL, archive_path)
+        fileobj = tarfile.open(archive_path, "r:gz").extractfile(
+            'CaliforniaHousing/cal_housing.data')
```
Would it be possible to use a `urlopen` instead of `urlretrieve` to read the content of `DATA_URL` in memory and pass it as a pseudo-fileobj to `tarfile.open`?

Alternatively wrap it as was done in master: `BytesIO(urlopen(DATA_URL).read())`. This should make it possible to avoid creating a temporary file for the archive.

lesteve committed:
Avoid downloading the tar.gz archive locally

ogrisel commented:
Looks great, thanks for the fix @lesteve. I will merge by squash and rebase.

ogrisel commented:
Done as 3daab8a99862674514735a3caadd45cc542b43c1
